drivers/sensor/bme680/bme680.c

```diff
@@ -0,0 +1,427 @@
+	if (calc_hum > 100000) { /* Cap at 100%rH */
+		calc_hum = 100000;
+	} else if (calc_hum < 0) {
+		calc_hum = 0;
+	}
+
+	data->calc_humidity = calc_hum;
+}
+
+static void bme680_calc_gas_resistance(struct bme680_data *data, u8_t gas_range,
+				       u16_t adc_gas_res)
+{
+	s64_t var1, var3;
+	u64_t var2;
+
+	u32_t look_up1[16] = { 2147483647, 2147483647, 2147483647, 2147483647,
+			       2147483647, 2126008810, 2147483647, 2130303777,
+			       2147483647, 2147483647, 2143188679, 2136746228,
+			       2147483647, 2126008810, 2147483647, 2147483647 };
+
+	u32_t look_up2[16] = { 4096000000, 2048000000, 1024000000, 512000000,
+			       255744255, 127110228, 64000000, 32258064,
+			       16016016, 8000000, 4000000, 2000000, 1000000,
+			       500000, 250000, 125000 };
+
+	var1 = (s64_t)((1340 + (5 * (s64_t)data->range_sw_err)) *
+		       ((s64_t)look_up1[gas_range])) >> 16;
+	var2 = (((s64_t)((s64_t)adc_gas_res << 15) - (s64_t)(16777216)) + var1);
+	var3 = (((s64_t)look_up2[gas_range] * (s64_t)var1) >> 9);
+	data->calc_gas_resistance = (u32_t)((var3 + ((s64_t)var2 >> 1))
+					    / (s64_t)var2);
+}
+
+static u8_t bme680_calc_res_heat(struct bme680_data *data, u16_t heatr_temp)
+{
+	u8_t heatr_res;
+	s32_t var1, var2, var3, var4, var5;
+	s32_t heatr_res_x100;
+	s32_t amb_temp = 25;    /* Assume ambient temperature to be 25 deg C */
+
+	if (heatr_temp > 400) { /* Cap temperature */
+		heatr_temp = 400;
+	}
+
+	var1 = ((amb_temp * data->par_gh3) / 1000) * 256;
+	var2 = (data->par_gh1 + 784) * (((((data->par_gh2 + 154009)
+					   * heatr_temp * 5) / 100)
+					 + 3276800) / 10);
+	var3 = var1 + (var2 / 2);
+	var4 = (var3 / (data->res_heat_range + 4));
+	var5 = (131 * data->res_heat_val) + 65536;
+	heatr_res_x100 = ((var4 / var5) - 250) * 34;
+	heatr_res = (heatr_res_x100 + 50) / 100;
+
+	return heatr_res;
+}
+
+static u8_t bme680_calc_gas_wait(u16_t dur)
+{
+	u8_t factor = 0, durval;
+
+	if (dur >= 0xfc0) {
+		durval = 0xff; /* Max duration*/
+	} else {
+		while (dur > 0x3F) {
+			dur = dur / 4;
+			factor += 1;
+		}
+		durval = dur + (factor * 64);
+	}
+
+	return durval;
+}
+
+static int bme680_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct bme680_data *data = dev->driver_data;
+	u8_t buff[BME680_LEN_FIELD] = { 0 };
+	u8_t gas_range;
+	u32_t adc_temp, adc_press;
+	u16_t adc_hum, adc_gas_res;
+	int size = BME680_LEN_FIELD;
+	int ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	ret = bme680_reg_read(data, BME680_REG_FIELD0, buff, size);
+	if (ret < 0) {
+		return ret;
+	}
+
+	data->new_data = buff[0] & BME680_MSK_NEW_DATA;
+	data->heatr_stab = buff[14] & BME680_MSK_HEATR_STAB;
+
+	adc_press = (u32_t)(((u32_t)buff[2] * 4096) | ((u32_t)buff[3] * 16)
+			    | ((u32_t)buff[4] / 16));
+	adc_temp = (u32_t)(((u32_t)buff[5] * 4096) | ((u32_t)buff[6] * 16)
+			   | ((u32_t)buff[7] / 16));
+	adc_hum = (u16_t)(((u32_t)buff[8] * 256) | (u32_t)buff[9]);
+	adc_gas_res = (u16_t)((u32_t)buff[13] * 4 | (((u32_t)buff[14]) / 64));
```

Comment:
Use logical shifts instead of multiply/divide by powers of two

Reply:
ok.